[MLIR][TOSA] Update IfOp print/parse to support ranked condition tens…

Yuvaraj-Venkatesh opened the pull request:
…or and optional block arguments

This change extends the TOSA `cond_if` operation's print and parse logic to handle the following:

- The condition operand may now have any rank, as long as the total number of elements sums to 1.

  %1 = tosa.cond_if %0 : tensor<1x1x1xi1> -> tensor<4xf32>

- The `then` and `else` regions can now include optional block arguments. The updated IR syntax reflects this:

  %1 = tosa.cond_if %0 (%arg2 = %arg0, %arg3 = %arg1) : tensor<i1> (tensor<4xf32>, tensor<4xf32>) -> tensor<4xf32>

- Removed parentheses around single result types in the printed representation, aligning with the `AsmPrinter` conventions.

Yuvaraj-Venkatesh committed:
[MLIR][TOSA] Update IfOp print/parse to support ranked condition tensor and optional block arguments

This change extends the TOSA `cond_if` operation's print and parse logic to handle the following:

- The condition operand may now have any rank, as long as the total number of elements sums to 1.

  %1 = tosa.cond_if %0 : tensor<1x1x1xi1> -> tensor<4xf32>

- The `then` and `else` regions can now include optional block arguments. The updated IR syntax reflects this:

  %1 = tosa.cond_if %0 (%arg2 = %arg0, %arg3 = %arg1) : tensor<i1> (tensor<4xf32>, tensor<4xf32>) -> tensor<4xf32>

- Removed parentheses around single result types in the printed representation, aligning with the `AsmPrinter` conventions.

Yuvaraj-Venkatesh commented on mlir/test/Dialect/Tosa/verifier.mlir:
```diff
@@ -500,9 +500,37 @@ func.func @test_cond_if_input_list_mismatch_else_block_2(%arg0: tensor<f32>, %ar
+func.func @test_cond_if_input_list_mismatch_else_block_simple(%arg0: tensor<f32>, %arg1: tensor<f32>, %arg2: tensor<i1>) -> tensor<f32> {
+  // expected-error@+1 {{'tosa.cond_if' op require same number of values in 'else_graph' arguments (1) and 'input_list' (2)}}
+  %0 = tosa.cond_if %arg2 (%arg3 = %arg0, %arg4 = %arg1) : tensor<i1> (tensor<f32>, tensor<f32>) -> tensor<f32> {
+    %1 = tosa.add %arg3, %arg4 : (tensor<f32>, tensor<f32>) -> tensor<f32>
```
@lhutton1  Thanks for the review.

I have modified the code to have block arguments in both then and else region.

Yuvaraj-Venkatesh commented on mlir/test/Dialect/Tosa/error_if_check.mlir:
```diff
@@ -280,3 +280,32 @@ func.func @test_cond_if_isolated_from_above(%arg0: tensor<f32>, %arg1: tensor<f3
+// -----
+
+// CHECK-LABEL: cond_if_cond_type
+func.func @test_cond_if_cond_type(%arg0: tensor<f32>, %arg1: tensor<f32>, %arg2: tensor<i1>) -> tensor<f32> {
```
Moved to verifier.mlir . Thanks

Yuvaraj-Venkatesh commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -3646,25 +3646,93 @@ std::optional<SmallVector<int64_t, 4>> ApplyScaleOp::getShapeForUnroll() {
+
+    OptionalParseResult typeResult = parser.parseOptionalType(functionType);
+    if (typeResult.has_value() && failed(typeResult.value()))
+      return parser.emitError(parser.getCurrentLocation(),
```
Added a test case in verifier.mlir

lhutton1 commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -3646,23 +3646,87 @@ std::optional<SmallVector<int64_t, 4>> ApplyScaleOp::getShapeForUnroll() {
+  if (listResult.has_value()) {
+    FunctionType functionType;
+
+    OptionalParseResult typeResult = parser.parseOptionalType(functionType);
```
Could this be simplified by requiring a type i.e. using `parseType`? I think later code in this block assumes functionType was parsed successfully, which can cause a crash when type information is not provided correctly

Yuvaraj-Venkatesh replied:
Make sense. Thanks for pointing that out. I've updated to use parseType instead of parseOptionalType.

lhutton1 commented on mlir/lib/Dialect/Tosa/IR/TosaOps.cpp:
```diff
@@ -3646,23 +3646,87 @@ std::optional<SmallVector<int64_t, 4>> ApplyScaleOp::getShapeForUnroll() {
+    OptionalParseResult typeResult = parser.parseOptionalType(functionType);
+    if (typeResult.has_value() && failed(typeResult.value()))
+      return parser.emitError(parser.getCurrentLocation(),
+                              "expected type for block args");
```
nit: perhaps "expected list of types for block arguments followed by arrow type list of return types" is a more descriptive message?

lhutton1 reviewed (commented):
Thanks for the updates @Yuvaraj-Venkatesh! Apologies, I took another look and picked up some things I missed the first time round. Could we also add some positive test cases to check the improved functionality?:
- One to check that the shape of the condition is retained after parse/print round-trip
- One to check that the block arguments are retained after parse/print round-trip

Since the cond_if validation check is now merged, it would be good to update this comment: https://github.com/llvm/llvm-project/pull/143772/files#diff-9094a3c264abb63100ab95e0e9e6d5c303b346ec9926b814e58b50ffaa1253e0R1235 as part of this change as well.

Since this change potentially impacts downstream users, tagging a few more people that might be interested: @sjarus @GeorgeARM @amirBish @udaya-ranga

Yuvaraj-Venkatesh commented:
> Thanks for the updates @Yuvaraj-Venkatesh! Apologies, I took another look and picked up some things I missed the first time round. Could we also add some positive test cases to check the improved functionality?:
> 
> * One to check that the shape of the condition is retained after parse/print round-trip
> * One to check that the block arguments are retained after parse/print round-trip
> 
> Since the cond_if validation check is now merged, it would be good to update this comment: https://github.com/llvm/llvm-project/pull/143772/files#diff-9094a3c264abb63100ab95e0e9e6d5c303b346ec9926b814e58b50ffaa1253e0R1235 as part of this change as well.
> 
> Since this change potentially impacts downstream users, tagging a few more people that might be interested: @sjarus @GeorgeARM @amirBish @udaya-ranga

Thanks @lhutton1!
I’ve added a couple of positive test cases and updated the comments in the validation logic as well.

lhutton1 reviewed (approved):
Thanks for the updates @Yuvaraj-Venkatesh, LGTM!

Yuvaraj-Venkatesh committed:
Merge branch 'main' into cond-if-print-parse-improvements

lhutton1 committed:
Merge branch 'main' into cond-if-print-parse-improvements
